File: sdks/java/src/main/java/org/byteveda/taskito/dashboard/api/Metrics.java
package org.byteveda.taskito.dashboard.api;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.byteveda.taskito.model.TaskMetric;

/**
 * Pure aggregation of raw {@link TaskMetric} rows into the dashboard's
 * per-task summary and time-bucketed series. Durations are wall-time
 * nanoseconds rendered as milliseconds; percentiles use the nearest-rank method.
 */
final class Metrics {
    private Metrics() {}

    /** Group by task → {count, success/failure counts, avg/min/max, p50/p95/p99} (all ms). */
    static Map<String, Object> aggregateByTask(List<TaskMetric> rows) {
        Map<String, List<TaskMetric>> byTask = new LinkedHashMap<>();
        for (TaskMetric row : rows) {
            byTask.computeIfAbsent(row.taskName, k -> new ArrayList<>()).add(row);
        }
        Map<String, Object> out = new LinkedHashMap<>();
        byTask.forEach((task, taskRows) -> out.put(task, summarise(taskRows)));
        return out;
    }

    /** Bucket rows by {@code recordedAt / bucketMs}, oldest first. */
    static List<Map<String, Object>> timeseries(List<TaskMetric> rows, long bucketMs) {
        long width = bucketMs > 0 ? bucketMs : 60_000;
        Map<Long, List<TaskMetric>> buckets = new TreeMap<>();
        for (TaskMetric row : rows) {
            long start = (row.recordedAt / width) * width;
            buckets.computeIfAbsent(start, k -> new ArrayList<>()).add(row);
        }
        List<Map<String, Object>> out = new ArrayList<>();
        buckets.forEach((start, bucketRows) -> {
            Map<String, Object> bucket = new LinkedHashMap<>();
            bucket.put("timestamp", start);
            bucket.put("count", bucketRows.size());
            bucket.put("success", countSucceeded(bucketRows));
            bucket.put("failure", bucketRows.size() - countSucceeded(bucketRows));
            double[] sorted = sortedDurations(bucketRows);
            bucket.put("avg_ms", average(sorted));
            bucket.put("p50_ms", percentile(sorted, 50));
            bucket.put("p95_ms", percentile(sorted, 95));
            bucket.put("p99_ms", percentile(sorted, 99));
            out.add(bucket);
        });
        return out;
    }

    private static Map<String, Object> summarise(List<TaskMetric> rows) {
        double[] sorted = sortedDurations(rows);
        int success = countSucceeded(rows);
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("count", rows.size());
        m.put("success_count", success);
        m.put("failure_count", rows.size() - success);
        m.put("avg_ms", average(sorted));
        m.put("p50_ms", percentile(sorted, 50));
        m.put("p95_ms", percentile(sorted, 95));
        m.put("p99_ms", percentile(sorted, 99));
        m.put("min_ms", sorted.length == 0 ? 0.0 : sorted[0]);
        m.put("max_ms", sorted.length == 0 ? 0.0 : sorted[sorted.length - 1]);
        return m;
    }

    private static int countSucceeded(List<TaskMetric> rows) {
        int count = 0;
        for (TaskMetric row : rows) {
            if (row.succeeded) {
                count++;
            }
        }
        return count;
    }

    private static double[] sortedDurations(List<TaskMetric> rows) {
        double[] durations = new double[rows.size()];
        for (int i = 0; i < rows.size(); i++) {
            durations[i] = rows.get(i).wallTimeNs / 1_000_000.0;
        }
        java.util.Arrays.sort(durations);
        return durations;
    }

    private static double average(double[] sorted) {
        if (sorted.length == 0) {
            return 0.0;
        }
        double sum = 0;
        for (double value : sorted) {
            sum += value;
        }
        return sum / sorted.length;
    }

    /** Nearest-rank percentile: index = ceil(p/100 * n) - 1. */
    private static double percentile(double[] sorted, int p) {
        if (sorted.length == 0) {
            return 0.0;
        }
        int index = (int) Math.ceil(p / 100.0 * sorted.length) - 1;
        index = Math.max(0, Math.min(sorted.length - 1, index));
        return sorted[index];
    }
}

File: sdks/node/src/dashboard/handlers/metrics.ts
// Aggregate the core's per-execution metric rows into the SPA's per-task and
// time-bucketed shapes. The core records a metric for every finished job.

import type { Metric } from "../../types";

export interface TaskMetrics {
  count: number;
  success_count: number;
  failure_count: number;
  avg_ms: number;
  p50_ms: number;
  p95_ms: number;
  p99_ms: number;
  min_ms: number;
  max_ms: number;
}

export interface TimeseriesBucket {
  timestamp: number;
  count: number;
  success: number;
  failure: number;
  avg_ms: number;
  p50_ms: number;
  p95_ms: number;
  p99_ms: number;
}

function avg(values: number[]): number {
  return values.length ? values.reduce((a, b) => a + b, 0) / values.length : 0;
}

/** Nearest-rank percentile over an ascending-sorted list. */
function pct(sorted: number[], p: number): number {
  if (sorted.length === 0) {
    return 0;
  }
  // Nearest-rank: ceil(p/100 * n) - 1, clamped into [0, n-1].
  const rank = Math.ceil((p / 100) * sorted.length) - 1;
  const index = Math.min(sorted.length - 1, Math.max(0, rank));
  return sorted[index] ?? 0;
}

export function aggregateByTask(rows: Metric[]): Record<string, TaskMetrics> {
  const durations = new Map<string, number[]>();
  const successes = new Map<string, number>();
  for (const row of rows) {
    const list = durations.get(row.taskName) ?? [];
    list.push(row.wallTimeNs / 1e6);
    durations.set(row.taskName, list);
    if (row.succeeded) {
      successes.set(row.taskName, (successes.get(row.taskName) ?? 0) + 1);
    }
  }

  const out: Record<string, TaskMetrics> = {};
  for (const [task, list] of durations) {
    list.sort((a, b) => a - b);
    const count = list.length;
    const success = successes.get(task) ?? 0;
    out[task] = {
      count,
      success_count: success,
      failure_count: count - success,
      avg_ms: avg(list),
      p50_ms: pct(list, 50),
      p95_ms: pct(list, 95),
      p99_ms: pct(list, 99),
      min_ms: list[0] ?? 0,
      max_ms: list[count - 1] ?? 0,
    };
  }
  return out;
}

export function bucketTimeseries(rows: Metric[], bucketMs: number): TimeseriesBucket[] {
  const buckets = new Map<number, { durations: number[]; success: number }>();
  for (const row of rows) {
    const ts = Math.floor(row.recordedAt / bucketMs) * bucketMs;
    const bucket = buckets.get(ts) ?? { durations: [], success: 0 };
    bucket.durations.push(row.wallTimeNs / 1e6);
    if (row.succeeded) {
      bucket.success += 1;
    }
    buckets.set(ts, bucket);
  }

  return [...buckets.entries()]
    .sort((a, b) => a[0] - b[0])
    .map(([timestamp, bucket]) => {
      const sorted = bucket.durations.sort((a, b) => a - b);
      return {
        timestamp,
        count: sorted.length,
        success: bucket.success,
        failure: sorted.length - bucket.success,
        avg_ms: avg(sorted),
        p50_ms: pct(sorted, 50),
        p95_ms: pct(sorted, 95),
        p99_ms: pct(sorted, 99),
      };
    });
}
